include/core/Coupling.hpp

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+struct Coupling
+{
+    ~Coupling() = default;
+    
+    // Read values from connected input, and write (possibly converted,
+    // interpolated, scaled, etc) values to connected output
+    virtual void transfer() = 0;
+};
+
+struct DirectCoupling : public Coupling
+{
+    DirectCoupling(shared_ptr<SourceVariable> from, shared_ptr<SinkVariable> to)
+        : from_(from)
+        , to_(to)
+    {
+        assert(from->discretization()->equals(to->discretization));
+
+        assert(from->element_dimensionality() == to->element_dimensionality());
+        // TODO: relax this to automatic unit conversion by scaling
+        assert(from->element_units() == to->element_units());
+    }
+
+    DirectCoupling(shared_ptr<SourceVariable> from, shared_ptr<SinkVariable> to,
+                   double scaling)
+        : from_(from)
+        , to_(to)
+        , scaling_(scaling)
+    { 
+        assert(from->discretization()->equals(to->discretization));
+
+        // Don't assert that dimensionality/units are common - the
+        // scaling should be set to address that
+    }
+
+    void transfer() override
+    {
+        // Short-circuit the case of shared storage
+        if (from_ == to_)
+            return;
+
+        span<double> src = from_.storage(), dst = to_.storage();
+        for (size_t i = 0; i < src.size(); ++i) {
+            dst[i] = scaling * src[i];
+        }
+    }
+
+private:
+    shared_ptr<SourceVariable> from_;
+    shared_ptr<SinkVariable> to_;
+    double scaling = 1.0;
+};
+
+struct LumpedPrecipitationRateCoupling : public DirectCoupling
+{
+    PrecipitationRateCoupling(shared_ptr<SourceVariable> from,
+                              shared_ptr<SinkVariable> to,
+                              std::chrono::duration<double> timestep,
+                              double area_sqkm)
+        : DirectCoupling(from, to, timestep*area_sqkm*unit_correction)
+    {
+        assert(from->element_dimensionality() == length/time);
+        assert(to->element_dimensionality() == length*length*length);
+    }
+};
+
+// Can handle subsetting, permutation, linear interpolation,
+// (potentially conservative) partitioning, and other conversions that
+// involve only linear combinations of the input to produce the output
+struct LinearTransferCoupling : public Coupling
+{
+    LinearTransferCoupling(shared_ptr<SourceVariable> from,
+                           shared_ptr<SinkVariable> to,
+                           SparseMatrix<double> transfer_weights)
+        : from_(from)
+        , to_(to)
+        , transfer_weights_(transfer_weights)
+    { }
+
+    void transfer() override
+    {
+        SpMV(transfer_weights_(), from_.storage(), to.storage());
+    }
+
+    shared_ptr<SourceVariable> from_;
+    shared_ptr<SinkVariable> to_;
+    SparseMatrix transfer_weights_;
```

Comment:
Given what's available on WCOSS, it may actually be easiest to define this from a NumPy type, since that will have a better-than-naive implementation